Fix flakey `ocean.task` test (major `TimeSet` bug)

mihails-strasuns-sociomantic opened the pull request:
We have observed two times that some CI runs fail with totally unrelated test failures coming from task code. Previous case was ~3 weeks ago in one of PRs and it has happened again on v3.x.x HEAD after merging green PR today. It looks like one of task tests has some very low chance of failure that is not foreseen.

Will need further investigation.

mihails-strasuns-sociomantic commented:
Typical stack trace:

```
core.exception.AssertError@./src/ocean/task/Task.d(321): Assertion failure
----------------
src/core/exception.d:420 _d_assert [0x116bd2b]
??:? void ocean.task.Task.__assert(int) [0xfe178f]
./src/ocean/task/Task.d:321 void ocean.task.Task.Task.resume() [0xfe0179]
./src/ocean/task/util/Timer.d:300 void ocean.task.util.Timer.registerResumeEvent(ocean.task.Task.Task, uint).__lambda4(ref ocean.task.util.Timer.EventData) [0xdb36a3]
./src/ocean/io/select/client/TimerSet.d:214 void ocean.io.select.client.TimerSet.TimerSet!(ocean.task.util.Timer.EventData).TimerSet.Event.timeout() [0xcd4d3f]
./src/ocean/time/timeout/model/ExpiryRegistrationBase.d:264 ocean.time.timeout.model.ITimeoutClient.ITimeoutClient ocean.time.timeout.model.ExpiryRegistrationBase.ExpiryRegistrationBase.timeout() [0x115d098]
./src/ocean/time/timeout/TimeoutManager.d:511 ulong ocean.time.timeout.TimeoutManager.TimeoutManagerBase.checkTimeouts(ulong, scope bool delegate(ocean.time.timeout.model.ITimeoutClient.ITimeoutClient)) [0xe346f1]
./src/ocean/time/timeout/TimeoutManager.d:444 ulong ocean.time.timeout.TimeoutManager.TimeoutManagerBase.checkTimeouts(scope bool delegate(ocean.time.timeout.model.ITimeoutClient.ITimeoutClient)) [0xe344ec]
./src/ocean/io/select/timeout/TimerEventTimeoutManager.d:99 bool ocean.io.select.timeout.TimerEventTimeoutManager.TimerEventTimeoutManager.TimerEvent.handle_(ulong) [0xdb9e34]
./src/ocean/io/select/client/TimerEvent.d:361 bool ocean.io.select.client.TimerEvent.ITimerEvent.handle(ocean.sys.Epoll.epoll_event_t.Event) [0xfff978]
./src/ocean/io/select/selector/SelectedKeysHandler.d:151 void ocean.io.select.selector.SelectedKeysHandler.SelectedKeysHandler.handleSelectedKey(ocean.sys.Epoll.epoll_event_t, scope void delegate(Exception)) [0x115535e]
./src/ocean/io/select/selector/SelectedKeysHandler.d:113 void ocean.io.select.selector.SelectedKeysHandler.SelectedKeysHandler.opCall(ocean.sys.Epoll.epoll_event_t[], scope void delegate(Exception)) [0x11552ab]
./src/ocean/io/select/EpollSelectDispatcher.d:785 uint ocean.io.select.EpollSelectDispatcher.EpollSelectDispatcher.select(bool) [0xe5206c]
./src/ocean/io/select/EpollSelectDispatcher.d:701 void ocean.io.select.EpollSelectDispatcher.EpollSelectDispatcher.eventLoop(scope bool delegate(), scope void delegate(Exception)) [0xe51e7e]
./src/ocean/task/Scheduler.d:568 void ocean.task.Scheduler.Scheduler.eventLoop() [0xe16b51]
./src/ocean/task/util/Timer.d:200 void ocean.task.util.Timer.__unittestL167_248() [0xdb30d5]
??:? void ocean.task.util.Timer.__modtest() [0xdb36f2]
./src/ocean/core/UnitTestRunner.d:644 ocean.core.UnitTestRunner.UnitTestRunner.Result ocean.core.UnitTestRunner.UnitTestRunner.timedTest(object.ModuleInfo*, out core.sys.posix.sys.time.timeval, ref char[]) [0xf92afe]
./src/ocean/core/UnitTestRunner.d:261 _D5ocean4core14UnitTestRunner14UnitTestRunner3runMFZ14__foreachbody2MFPS6object10ModuleInfoZi [0xf91ac1]
src/object.d:1585 int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)).__lambda2(immutable(object.ModuleInfo*)) [0x116b257]
src/rt/minfo.d:289 int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))).__foreachbody2(ref rt.sections_elf_shared.DSO) [0x1174f6a]
src/rt/sections_elf_shared.d:53 int rt.sections_elf_shared.DSO.opApply(scope int delegate(ref rt.sections_elf_shared.DSO)) [0x1175435]
src/rt/minfo.d:282 int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))) [0x1174efb]
src/object.d:1585 int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)) [0x116b233]
./src/ocean/core/UnitTestRunner.d:205 int ocean.core.UnitTestRunner.UnitTestRunner.run() [0xf915ef]
./src/ocean/core/UnitTestRunner.d:947 _Dmain [0xf93637]
```

mihails-strasuns-sociomantic commented:
Added 93ab2c873c750fa1fedd0b44348442478f500e4c to get better trace next time it happens

nemanja-boric-sociomantic commented:
Judging by the backtrace, this seems to be the flaky test: https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/task/util/Timer.d#L167-L204

mihails-strasuns-sociomantic commented:
Got this again after addition of https://github.com/sociomantic-tsunami/ocean/commit/93ab2c873c750fa1fedd0b44348442478f500e4c and it implies something very weird.

https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/task/util/Timer.d#L286 doesn't fail but by the time of https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/task/util/Timer.d#L302 `event.to_resume` is null. @gavin-norman-sociomantic do you have any idea if this is even possible in absence of some memory corruption?

nemanja-boric-sociomantic commented:
Are we sure we're keeping the reference in the area available for the GC to scan? Looking here: https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/io/select/client/TimerSet.d#L327 we get one element from the pool which will store our `EventData`. We return it back from `task.Timer.registerResumeEvent`, but we don't store the reference to it: https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/task/util/Timer.d#L63 (only to some of its fields).

Now, if we had a bug inside our `ObjectPool` (or if it was the way it's designed) only reference would be kept inside `ExpiryRegistration` which is implemented using ebtree, so it's not scanned by GC, meaning triggering GC and reclaiming this memory would cause behaviour like this.

nemanja-boric-sociomantic commented:
It does look like `ObjectPool` keeps the references to the busy items (`IPool` implementation is just a single array with the border between busy and idle elements), but it's worth double-checking.

mihails-strasuns-sociomantic commented:
Also premature collection would result in broken but non-null pointer.

gavin-norman-sociomantic commented:
>@gavin-norman-sociomantic do you have any idea if this is even possible in absence of some memory corruption?

Sounds highly abnormal to me.

mihails-strasuns-sociomantic commented:
Also worth noting that so far I was not able to reproduce it locally even once, running all tests 1000 times in a row (with different `nice` value for a each thousand)

mihails-strasuns-sociomantic commented:
It looks like origin of the problem comes from here https://github.com/sociomantic-tsunami/ocean/commit/43e881d96f5a7e760bf8427266e3d9e0de371672#diff-cdfe73f7a32941d6ef2be1802a4a7819R680

mihails-strasuns-sociomantic commented:
But it is only manifestation of real problem - `terminationHook` system is broken when several delegates with same pointer are registered. In this specific case same hooks get added from both `wait` and `awaitOrTimeout` for the same task.

It will probably be more clear once I come up with fixup PR :)

mihails-strasuns-sociomantic commented:
On a related topic, @gavin-norman-sociomantic could you explain me something about `TimerSet` internals please? It looks like normally event is recycled from this method - https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/io/select/client/TimerSet.d#L212-L216 . But what happens when `unregister` is called instead (https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/io/select/client/TimerSet.d#L199-L202)?

mihails-strasuns-sociomantic committed:
Implement `drop` method for ExpiryRegistrationBase

New method, `drop`, provides same functionality as `unregister`, but also
removes the expiry from current list of fired events if it was present there.
This method makes possible to disable one expiry from timeout callback of
another expiry, even if they fire in the same time slot.

mihails-strasuns-sociomantic committed:
Fix concurrent unregistration of TimeSet events

Uses new `drop` method of expiry instead of `unregister` so that
events gets completely disabled even if it has already fired for current
time slot.

mihails-strasuns-sociomantic committed:
Fix memory leak in TimerSet event pool

TimerSet events were not recycled if event was unregistered before
firing.

mihails-strasuns-sociomantic commented:
Thanks to some very productive brain storming session with @nemanja-boric-sociomantic , root bug is identified as the fix is attached here :)

mathias-lang-sociomantic commented on relnotes/expiry.feature.md:
```diff
@@ -0,0 +1,6 @@
+  New method, `drop`, provides same functionality as `unregister`, but also
+  removes the expiry from current list of fired events if it was present there.
+  This method makes possible to disable one expiry from timeout callback of
+  another expiry, even if they fire in the same time slot.
```
This is a patch release though

mihails-strasuns-sociomantic replied:
Yes, this is one of rare exception cases "have to add feature to fix critical bug" in my opinion. I would probably even make the method `package(ocean)` protected if I could, to stick with formalities. But I can't, so better at least document it.

gavin-norman-sociomantic replied:
"even if they fire in the same time slot" -- here too. (Sorry to be pedantic, but I think this phrase adds unnecessary confusion.)

nemanja-boric-sociomantic commented on src/ocean/time/timeout/TimeoutManager.d:
```diff
@@ -228,6 +227,37 @@ abstract class TimeoutManagerBase : ITimeoutManager
+
+        ***********************************************************************/
+
+        protected void drop ( ref Expiry expiry )
```
Only thing I wonder about is how much is `unregister` useful now. Should anything call `unregister` now?

mihails-strasuns-sociomantic replied:
Yes, it is the very same reason why I couldn't just change `unregister` to do the same thing : https://github.com/mihails-strasuns-sociomantic/ocean/blob/a3b792b1dc6d1b0a87897fa9c5647467815f9b79/src/ocean/time/timeout/TimeoutManager.d#L524-L531

nemanja-boric-sociomantic replied:
Yes, now I remember. Feels like `unregister` should be private then, but I don't know how could we pull that off.

mihails-strasuns-sociomantic replied:
I don't think so as it has to be used across ocean modules. `package(ocean)` would fix it but it is not available in D1.

nemanja-boric-sociomantic replied:
Yeah. Oh well, such is life.

gavin-norman-sociomantic replied:
Should we make `unregister` private, in a future release? (Either as a breaking change or via a deprecation path.)

nemanja-boric-sociomantic replied:
I think this answers your last question @mihails-strasuns-sociomantic :-)

nemanja-boric-sociomantic replied:
Oh, yes, I asked the same question @gavin-norman-sociomantic , but unregister can't be private, since it needs to be accessible from other ocean internal modules - `package` would be appropriate, but it doesn't work.

mihails-strasuns-sociomantic replied:
Maybe we can do some trick with `protected`, but I'd delay that for v3.x.x

nemanja-boric-sociomantic commented on src/ocean/time/timeout/TimeoutManager.d:
```diff
@@ -228,6 +227,37 @@ abstract class TimeoutManagerBase : ITimeoutManager
+            foreach (ref pending; this.outer.expired_registrations[])
+            {
+                if (pending is registration)
+                    pending = null;
```
According to the test coverage, this line is red :-). I wonder if it we merge it and suddenly becomes green :-).

Also, I wonder what about "files not in diff unexpectedly changed coverage". I hope codecov has some threshold for this.

mihails-strasuns-sociomantic replied:
The task test which uncovered this was only added in 3.5.0 and this PR targets 3.4.x - I decided that adding yet another one just for the purpose of more coverage in old branch is not rational :)

gavin-norman-sociomantic commented on src/ocean/task/Scheduler.d:
```diff
@@ -568,6 +568,8 @@ final class Scheduler
         assert (this.state != State.Shutdown);
         this.state = State.Running;
 
+        debug_trace("Starting `theScheduler.eventLoop`");
```
Bug fix?

gavin-norman-sociomantic commented on src/ocean/time/timeout/TimeoutManager.d:
```diff
@@ -228,6 +227,37 @@ abstract class TimeoutManagerBase : ITimeoutManager
 
         /***********************************************************************
 
+            Same as `unregister` but also removes expirty from list of currently
```
expirty

gavin-norman-sociomantic replied:
I'd make this more specific:

"
Unregisters the specified expiry. If the expiry is present in the list of expired registrations being currently iterated over by `checkTimeouts`, then it will be removed (its `timeout` method will not be called). (This means that `drop` can be called from `timeout` callbacks, unlike `unregister`.)

If the specified expiry is currently not registered, nothing is done.
"

I'm not sure if this part of the doc for `unregister` is also relevant for `drop`:
>The subclass may reset the client instance to null after it has called this method (even if it throws an exception).

mihails-strasuns-sociomantic replied:
> I'm not sure if this part of the doc

I have no idea what it does even mean

gavin-norman-sociomantic replied:
Me neither. Maybe leave it off ;)

gavin-norman-sociomantic commented on src/ocean/time/timeout/TimeoutManager.d:
```diff
@@ -228,6 +227,37 @@ abstract class TimeoutManagerBase : ITimeoutManager
+
+        protected void drop ( ref Expiry expiry )
+        {
+            // If this expiry is already present in the expired registration
```
This is a bit clearer and more specific:

"If this method is called while `checkTimeouts` is iterating over the list of expired registrations, then we need to check whether the expiry to be dropped is present in the list and remove it, if it is. This makes it possible to disable timer events from the timeout callbacks of other timer events, even if they compete for the same time slot."

>even if they compete for the same time slot

What does this mean, by the way?

mihails-strasuns-sociomantic replied:
> What does this mean, by the way?

It was a shortest intuitive explanation of how `TimeoutManager` works I could think of :) i.e. that it fires once for actual epoll timer but multiplexes to several timer events for that one "slot".

gavin-norman-sociomantic replied:
Ah I see. It seems pretty redundant to me, as timeout callbacks are always called from `checkTimeouts`. I'd just leave it as:
"This makes it possible to disable timer events from the timeout callbacks of other timer events."

mihails-strasuns-sociomantic replied:
Updated

mihails-strasuns-sociomantic commented:
Rebased

gavin-norman-sociomantic commented on src/ocean/time/timeout/TimeoutManager.d:
```diff
@@ -228,6 +227,37 @@ abstract class TimeoutManagerBase : ITimeoutManager
+                *this.outer.expiry_to_client.get(&expiry);
+            foreach (ref pending; this.outer.expired_registrations[])
+            {
+                if (pending is registration)
```
Could you add a comment on the `this.outer.expired_registrations` field stating that the elements may be set to null by this method?

gavin-norman-sociomantic commented on src/ocean/time/timeout/TimeoutManager.d:
```diff
@@ -513,6 +543,9 @@ abstract class TimeoutManagerBase : ITimeoutManager
         // the optional delegate returns false.
         foreach (ref registration; this.expired_registrations[])
         {
+            if (registration is null)
```
Maybe add a comment: `// May be null if removed by drop`.

gavin-norman-sociomantic commented on src/ocean/time/timeout/model/ExpiryRegistrationBase.d:
```diff
@@ -96,6 +96,20 @@ abstract class ExpiryRegistrationBase : IExpiryRegistration
 
         /***********************************************************************
 
+            Same as `unregister` but also removes expirty from list of currently
```
Same comment suggestion as above.

gavin-norman-sociomantic commented on src/ocean/time/timeout/model/ExpiryRegistrationBase.d:
```diff
@@ -96,6 +96,20 @@ abstract class ExpiryRegistrationBase : IExpiryRegistration
 
+            Same as `unregister` but also removes expirty from list of currently
+            expired registrations. That means it can be called from `timeout`
+            callbacks, contrary to `unregister`.
```
Could you extend the doc of `update` to explain why it should not be called from a `timeout` callback?

Also: should `unregister` be called at all by any non-library code any more? (If not, please document this as well.)

mihails-strasuns-sociomantic replied:
> should unregister be called at all by any non-library code any more?

I was hoping you can answer that :) To me it seems 100% wrong to call it with old semantics but I don't have a good overview of how it is used across apps.

gavin-norman-sociomantic commented on src/ocean/time/timeout/model/ExpiryRegistrationBase.d:
```diff
@@ -190,22 +204,23 @@ abstract class ExpiryRegistrationBase : IExpiryRegistration
-        {
-            return false;
-        }
+        Same as `unregister` but also removes expirty from list of currently
```
Same comment suggestion.

gavin-norman-sociomantic commented on src/ocean/time/timeout/model/ExpiryRegistrationBase.d:
```diff
@@ -347,6 +362,44 @@ abstract class ExpiryRegistrationBase : IExpiryRegistration
     /***************************************************************************
 
+        Extracted common code of `drop` and `unregister`.
+
```
Please document what the extracted logic does.

gavin-norman-sociomantic commented on src/ocean/io/select/client/TimerSet.d:
```diff
@@ -202,6 +202,7 @@ public class TimerSet ( EventData ) : TimerEventTimeoutManager
         public void unregister ( )
         {
             this.expiry_registration.drop();
+            this.outer.events.recycle(this);
```
Good catch!

mihails-strasuns-sociomantic commented:
Updated:
- documentation changes as suggested by @gavin-norman-sociomantic 
- removed unnecessary commit with extra scheduler traces

gavin-norman-sociomantic commented on src/ocean/time/timeout/TimeoutManager.d:
```diff
@@ -228,6 +230,41 @@ abstract class TimeoutManagerBase : ITimeoutManager
+            // the expiry to be dropped is present in the list and remove it, if
+            // it is. This makes it possible to disable timer events from the
+            // timeout callbacks of other timer
+            // events, even if they compete for the same time slot.
```
Ditto about leaving off the bit about competing for a time slot.

mihails-strasuns-sociomantic commented:
Fixed relnotes too
